### ISSUE TITLE
video library not working on debian 12: "Unable to load library 'avresample'" and other error messages

### ISSUE THREAD

**Issue by joachimheintz** [processing/processing-video#221]
i cannot use the video library at all.
this is my system (uname -a):
`Linux debian12 6.1.0-12-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.52-1 (2023-09-07) x86_64 GNU/Linux`
and java --version reports:
```
openjdk 17.0.8 2023-07-18
OpenJDK Runtime Environment (build 17.0.8+7-Debian-1deb12u1)
OpenJDK 64-Bit Server VM (build 17.0.8+7-Debian-1deb12u1, mixed mode, sharing)
```
and this is what i did:
1. downloaded processing 4-3 from https://github.com/processing/processing4/releases/download/processing-1293-4.3/processing-4.3-linux-x64.tgz, unpacked and ran.
2. installed Video Library for Processing 4 via Sketch > Import Library > Manage Libraries.
3. ran _AsciiVideo_ from the Examples (File > Examples > Libraries > Video Library for Processing 4 > Capture > AsciiVideo)

the example does not run, reporting a link error for libavresample (and at the end an error about libsoup which i don't understand). i did some search and followed the advice in #214 to create a symlink.  but this did not help.

none of the examples for the video library work.  all fail with the error message below.  otherwise processing works allright.
```
can't load library avresample (avresample|libavresample|libavresample-0) with -Djna.library.path=/home/jh/sketchbook/libraries/video/library/linux-amd64. Last error:java.lang.UnsatisfiedLinkError: Unable to load library 'avresample':
/home/jh/sketchbook/libraries/video/library/linux-amd64/libavresample.so: file too short
/home/jh/sketchbook/libraries/video/library/linux-amd64/libavresample.so: file too short
libavutil.so.56: cannot open shared object file: No such file or directory
Native library (linux-x86-64/libavresample.so) not found in resource path (/tmp/processing/AsciiVideo11439244291195431793temp:/home/jh/processing-4.3/core/library/core.jar:/home/jh/processing-4.3/core/library/gluegen-rt.jar:/home/jh/processing-4.3/core/library/jogl-all.jar:/home/jh/sketchbook/libraries/video/library/gst1-java-core-1.4.0.jar:/home/jh/sketchbook/libraries/video/library/jna.jar:/home/jh/sketchbook/libraries/video/library/video.jar)

(process:55555): libsoup-ERROR **: 18:49:04.110: libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.
Could not run the sketch (Target VM failed to initialize).
For more information, read Help → Troubleshooting.
```

**Comment by bewegende-Architektur** [comment]
Hallo,

the same error occurs on Manjaro Linux:
> 6.6.10-1-MANJARO #1 SMP PREEMPT_DYNAMIC Fri Jan  5 17:38:36 UTC 2024 x86_64 GNU/Linux

> (process:103631): libsoup-ERROR **: 15:43:09.232: libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.
> Could not run the sketch (Target VM failed to initialize).
> For more information, read Help → Troubleshooting.

Is there an update on this issue?
Thanks in advance.

Greetings

**Comment by kflak** [comment]
hi,

same thing here on Arch Linux: 
```
(process:291408): libsoup-ERROR **: 11:43:43.260: libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.
Could not run the sketch (Target VM failed to initialize).
For more information, read Help → Troubleshooting.
```
java 21

**Comment by fukuchi** [comment]
Hi,

I have found a workaround. The problem is caused by the library files that come with the official release of processing-video. Removing them may solve the problem, as Processing will use the equivalent libraries installed system-wide.

For example, if your sketchbook folder is ~/sketchbook/, the libraries are located in ~/sketchbook/libraries/video/library/linux-amd64. Delete or rename the folder and restart Processing.

You need to install the equivalent libraries system-wide by using apt. You should probably install the following:

- libgstreamer1.0-0
- gstreamer1.0-plugins-good

**Comment by kflak** [comment]
Thanks @fukuchi! Unfortunately this doesn't work on my arch linux installation. I don't have the time now to plough through the errors, but I'm pretty sure I have the dependencies installed. Will look into it a bit more when I have time. If I remember correctly the gstreamer libraries where packaged together with processing a while ago when gstreamer broke openframeworks and processing with a new version. 1.20 if my memory serves me right...

**Comment by fukuchi** [comment]
@kflak I forgot to write about my environment. I am using Ubuntu 23.10 and the installed gstreamer seems to be version 1.22.

**Comment by kflak** [comment]
Yeah, I guess I'm probably missing some systemwide libraries... Will try to find them on Arch. It does seem like Processing is looking for them but can't find them when I delete the linux-arm libs.

**Comment by kflak** [comment]
Wow, thanks a lot for the workaround! I can confirm that reverting back to an old version of the video library I happened to have lying around did the trick. In my case I wasn't able to install processing 3.5, as it doesn't play well with my particular java environment, but video lib 2.0 does indeed seem to do the trick.

**Comment by trackme518** [comment]
The problem persists on` Ubuntu 24.0.4 LTS` and `Processing 4.2`.
`libsoup3 symbols detected. Using libsoup2 and libsoup3 in the same process is not supported.`
You could do` apt-get remove -y libsoup2.4` but other programs rely on that so that is NOT a solution.

The solution is just using the soup module from GStreamer 1.20 or newer as it detects which version of libsoup is in use and only uses that one. Earlier versions could only be built against a single version, thus breaking anything mixing the other version. So I am calling to maintainers to include that module in the GStreamer build for Processing library.

**Comment by tirigall** [comment]
I had the same problem on Ubuntu 24.04. I was able to fix it by renaming the `gstreamer-1.0` folder to `_gstreamer-1.0`.
The folder is located at: `Processing/libraries/video/library/linux-amd64/`
This way the program will use the Linux gstreamer instead of the Processing one.

**Comment by joshgiesbrecht** [comment]
Having tested various approaches, here's my recommendation:
Just re-release the current processing-video lib without GStreamer included for Linux.

I've tested using an updated GStreamer build, and it improves things but still has problems. After discussion on our Discord, it's also become apparent that GStreamer is a core Gnome library and we can just assume the OS has it included with little or no issues for end users. (If anyone submits problems because somehow their distro doesn't include it by default, we could add an error message letting people know to install a gstreamer package for their OS.)

Allowing Processing to fall back to the present GStreamer library on the system, on the other hand works without any apparent flaws at all on my system. Given that this is the workaround Linux users have been using for literally years, I'm pretty sure it's not just me.

**The only catch**:
I can't submit a PR for this, because nothing in the current processing-library build setup actively pulls in GStreamer or breaks if it's not there for Linux. A maintainer just needs to rerun the 'dist' build with the library/linux-amd64/ folder empty rather than manually populating it with GStreamer first. (I assume that any Windows / MacOS GStreamer files should still be kept in their folders as usual.)
